lib/kramdown/converter/math_engine/sskatex/details.rb

```diff
@@ -0,0 +1,264 @@
+    JS_ESCAPE = {
+      "\\" => "\\\\",
+      "\"" => "\\\"",
+      # Escaping single quotes not necessary in double-quoted string literals
+      #"'" => "\\'",
+      # JS does not recognize \a nor GNU's \e
+      # \b is ambiguous in regexps, as in Perl and Ruby
+      #"\b" => "\\b",
+      "\f" => "\\f",
+      "\n" => "\\n",
+      "\r" => "\\r",
+      "\t" => "\\t",
+      "\v" => "\\v",
+    }
+
+    # Turn a string into an equivalent Javascript literal, double-quoted.
+    # Similar to +.to_json+, but escape all non-ASCII codes as well.
+    def jsquote(str)
+      # No portable way of escaping Unicode points above 0xFFFF
+      '"%s"' % str.gsub(/[\0-\u{001F}""\\\u{0080}-\u{FFFF}]/u) { |c|
+        JS_ESCAPE[c] || "\\u%04x" % c.ord
+      }
+    end
+
+    # This should really be provided by +ExecJS::Runtimes+:
+    # A list of available JS engines, as symbols, in the order of preference.
+    def jsruns_available
+      ExecJS::Runtimes.runtimes.select(&:available?).map(&JSRUN_TOSYM)
+    end
+
+    # Given a +math_engine_opts+ dictionary,
+    # return a new JS engine context, initialized with the JS helper files,
+    # and with a JS object +katexopts+ containing general KaTeX options.
+    def newcontext(config)
+      # Configure JS engine. If specified, config[ :jsrun ] takes precedence,
+      # followed by an equivalent of ExecJS::Runtimes.autodetect,
+      # i.e. ENV_EXECJS_RUNTIME (parsed in a less adventurous way)
+      # or else the best available engine.
+      # If the result is invalid, unknown, or unavailable, errors will ensue.
+      jsrun = (config[:jsrun] ||
+               ENV_EXECJS_RUNTIME ||
+               JSRUN_TOSYM[ExecJS::Runtimes.best_available] ||
+               'Disabled').to_s.to_sym
+      ExecJS.runtime = JSRUN_FROMSYM[jsrun]
+      if config[:verbose]
+        warn "Available JS runtimes: #{jsruns_available.join(', ')}"
+        warn "Selected JS runtime: #{jsrun}"
+      end
+      # JS files and locations
+      jslib = config[:jslib] || DEFAULT_JSDIR
+      katexjs = config[:katexjs] || DEFAULT_KATEXJS
+      libs = config[:libs] || DEFAULT_LIBS
+      # ExecJS.compile is not incremental, so we have to concatenate sources
+      js = ''
+      libs.each do |libfile|
+        absfile = File.expand_path(libfile, jslib)
+        warn "Loading JS file: #{absfile}" if config[:verbose]
+        js << IO.read(absfile, external_encoding: Encoding::UTF_8) << "\n"
+      end
+      warn "Loading KaTeX JS file: #{katexjs}" if config[:verbose]
+      js << IO.read(katexjs, external_encoding: Encoding::UTF_8) << "\n"
+      # Initialize JS variable katexopts
+      katexopts = config[:katexopts] || DEFAULT_KATEXOPTS
+      js << "katexopts = #{katexopts.to_json}\n"
+      ExecJS.compile(js)
+    end
+
+    # Return a properly initialized JS engine context,
+    # depending on the given _config_. Cache its reference for reuse.
+    def jscontext(config)
+      JSCTX[config] ||= newcontext(config)
```

Comment:
I think that this could potentially be used for a denial of service attack since if kramdown is used by server application and parses client input, a client could modify `math_engine_opts` continuously to create new entries here.

How much memory does one entry take? How much time overhead is this compilation step?

Reply:
Caching the context is necessary because the MathEngine API does not seem to provide a method for just the engine initialization and processing of options. I have no figures yet, but I remember that the speedup compared to re-initialization for each formula was substantial enough that I would not go back.

In a typical Jekyll setup, `kramdown: math_engine_opts:` are configured globally (in `_config.yaml`), so the context initialization gets done only once per `jekyll build` (or `jekyll serve`). Therefore it is shared across all kramdown source files, which is particularly beneficial for large sites.

Regarding overall security: If the user controls part of the filesystem (e.g. can upload files) and can tweak `math_engine_opts` without restriction, the user could point `katexjs` (or any of the `libs`) to a file containing a definition of `katex.renderToString` (or our entry function `tex_to_html`)  with an endless loop or more nefarious intent. The Javascript in there could do anything, with the privileges of the kramdown process. So, unless user-tweakable kramdown options can be restricted for such a service, the gates to hell are open. What thread model do we have?

Attack scenarios aside: Indeed a long-running process should have a way of forgetting no-longer used configs, otherwise memory usage would be unbounded.
A minimalistic approach would be to cache only the last used config. This already works for Jekyll.
A better approach would be to clear the context cache when it takes more than a prescribed limit of memory. I'd work in that direction.

Reply:
Some benchmarks for a small site with 178 formulae, JS engine Duktape and default options:
  * 0.41s for `newcontext`
  * avg 0.018s per `compile_tex_math` (excluding `jscontext`)
  * Many inline formula take less than 3ms.

Therefore, caching at least one context is necessary.

Reply:
It occurs to me that recursive computation of context object sizes would be mostly unnecessary overhead. Instead, the configuration cache could be cleared if the number of its key-value pairs is about to exceed some maximum, say 10.
I'd even make that limit configurable via `math_engine_opts` because not doing so does not really improve security as long as users can (and need to) specify arbitrary `katexjs`.

Reply:
Thanks for investigating!

Since caching is necessary, instead of simple hash, it would probably be better than to use an LRU data structure, like this https://github.com/gettalong/hexapdf/blob/master/lib/hexapdf/utils/lru_cache.rb.

As for making the number of cached items configurable: Since the `math_engine_opts` can change with every input document, I don't see the use of making this configurable. I'd say that caching the last 10 context objects would just be fine.

Reply:
> Since caching is necessary, instead of simple hash, it would probably be better than to use an LRU data structure, like this https://github.com/gettalong/hexapdf/blob/master/lib/hexapdf/utils/lru_cache.rb.

Cool. That one is Affero GPL, so I suppose that code needs a separate notice in COPYING, unless you, the author, decide otherwise for kramdown.

Reply:
> I don't see the use of making this configurable. I'd say that caching the last 10 context objects would just be fine.

Agreed because making that configurable exposes an implementation detail which we may want to change without having to change the user documentation.